### PR TITLE
Mmr client gadget - support pallet reset

### PR DESCRIPTION
**Pull request by serban300**
Prerequisite for fixing: #12864

### CODE REVIEW

**Review by svyatonik — APPROVED** [review]
LGTM!

File: client/merkle-mountain-range/src/offchain_mmr.rs [hunk 1]
```diff
@@ -231,10 +239,22 @@ where
...
 			}
-			if let Err(e) =
-				aux_schema::write_gadget_state::<B, BE>(&*self.backend, &self.best_canonicalized)
-			{
-				debug!(target: LOG_TARGET, "error saving state: {:?}", e);
+			self.write_gadget_state_or_log();
+		}
+	}
+
+	fn update_first_mmr_block(&mut self, notification: &FinalityNotification<B>) {
```
Comment (acatangiu) [inline client/merkle-mountain-range/src/offchain_mmr.rs:246]: nit:
```suggestion
	fn handle_potential_pallet_reset(&mut self, notification: &FinalityNotification<B>) {
```

**Review by acatangiu — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2186198

File: client/merkle-mountain-range/src/aux_schema.rs [hunk 2]
```diff
@@ -45,7 +45,7 @@ pub(crate) fn write_gadget_state<B: Block, BE: AuxStore>(
 	backend.insert_aux(&[(GADGET_STATE, state.encode().as_slice())], &[])
 }
 
-fn load_decode<B: AuxStore, T: Decode>(backend: &B, key: &[u8]) -> ClientResult<Option<T>> {
+fn load_typed<B: AuxStore, T: Decode>(backend: &B, key: &[u8]) -> ClientResult<Option<T>> {
```
[the remark this answers is not in the document]
Comment (serban300) [reply client/merkle-mountain-range/src/aux_schema.rs]: Ok, I see. Changed back to the old name.

**Comment by serban300** [comment]
bot merge

**Pull request merged**